lasagne/layers/autoenc.py

```diff
@@ -0,0 +1,262 @@
+        The :class:`Layer` instance with respect to which a symmetric
+        autoencoder is built.
+    nonlinearity : 'same', list, callable, or None
+        The nonlinearity that is applied to the layer activations of theano
+        :class:`InverseLayer`. If None is provided, the layer will be linear.
+        If 'same', the :class:`InverseLayer` has the same nonlinearity as the
+        mirror :class:`Layer`.
+    b : 'new', list, Theano shared variable, numpy array, callable or None
+        An initializer for the biases. If 'new', new biases are initialized
+        using the default initialization. If a shared variable or a numpy array
+        is provided, the shape must match the incoming shape, skipping those
+        axes the biases are shared over (see below for an example). If set to
+        ``None``, the layer will have no biases and pass through its input
+        unchanged.
+        See :func:`lasagne.utils.create_param` for more information.
+
+    Returns
+    -------
+    layer: :class:`Layer` instance
+       The output :class:`Layer` of the symmetric autoencoder with
+       tied weights.
+
+    Examples
+    --------
+    >>> import lasagne
+    >>> from lasagne.layers import InputLayer, DenseLayer
+    >>> from lasagne.layers import build_autoencoder
+    >>> l_in = InputLayer((100, 20))
+    >>> l1 = DenseLayer(l_in, num_units=50)
+    >>> l2 = DenseLayer(l1, num_units=10)
+    >>> l_ae = build_autoencoder(l2, nonlinearity='same', b=None)
+    """
+
+    if isinstance(nonlinearity, (tuple, list)):
+        n_idx = 0
+
+    if isinstance(b, (tuple, list)):
+        b_idx = 0
+
+    layer = unfold_bias_and_nonlinearity_layers(layer)
+    layers = get_all_layers(layer)
+    autoencoder_layers = [layer]
+
+    for layer in layers[::-1]:
+        if not isinstance(layer, InputLayer):
+            incoming = autoencoder_layers[-1]
+
+            if isinstance(layer, NonlinearityLayer):
+                if nonlinearity == 'same':
+                    kwargs = dict(nonlinearity=layer.nonlinearity)
+                elif nonlinearity is None:
+                    kwargs = dict(nonlinearity=nonlinearities.identity)
+                elif isinstance(nonlinearity, (tuple, list)):
+                    kwargs = dict(nonlinearity=nonlinearity[n_idx])
+                    n_idx += 1
+                else:
+                    kwargs = dict(nonlinearity=nonlinearity)
+                a_layer = NonlinearityLayer(
+                    incoming=incoming,
+                    **kwargs
+                )
+            elif isinstance(layer, DropoutLayer):
+                a_layer = DropoutLayer(
+                    incoming=incoming,
+                    p=layer.p,
+                    rescale=layer.rescale
+                )
+            elif isinstance(layer, GaussianNoiseLayer):
+                a_layer = GaussianNoiseLayer(
+                    incoming=incoming,
+                    sigma=layer.sigma
+                )
```

Comment:
That's going to be hard to maintain. Whenever somebody adds another noise layer, we'll have to remember to add it here as well. Maybe we should give layers an optional `copy(incoming=...)` method that just recreates another instance of the same layer with the same parameters and a different incoming layer? All layers that don't have a `copy` method would be handled by the general `else:` case below. What would our design goals tell us to do? Of course this would just slightly change things from "having to think about adding it here" to "having to think about adding a copy() method", but the latter would be quite obvious when starting with an existing noise layer.
Hmm... `copy()` actually is a misleading name, `decoder()` would be better... but then we'd start putting all this functionality into the class hierarchy... that would be a lot cleaner, but too obtrusive. Those pesky design choices!